go/vt/vtctl/reparentutil/util.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+	"vitess.io/vitess/go/vt/topotools"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// ChooseNewPrimary finds a tablet that should become a primary after reparent.
+// The criteria for the new primary-elect are (preferably) to be in the same
+// cell as the current primary, and to be different from avoidPrimaryAlias. The
+// tablet with the most advanced replication position is chosen to minimize the
+// amount of time spent catching up with the current primary.
+//
+// Note that the search for the most advanced replication position will race
+// with transactions being executed on the current primary, so when all tablets
+// are at roughly the same position, then the choice of new primary-elect will
+// be somewhat unpredictable.
+func ChooseNewPrimary(
+	ctx context.Context,
+	tmc tmclient.TabletManagerClient,
+	shardInfo *topo.ShardInfo,
+	tabletMap map[string]*topo.TabletInfo,
+	avoidPrimaryAlias *topodatapb.TabletAlias,
+	waitReplicasTimeout time.Duration,
+	// (TODO:@ajm188) it's a little gross we need to pass this, maybe embed in the context?
+	logger logutil.Logger,
+) (*topodatapb.TabletAlias, error) {
+	if avoidPrimaryAlias == nil {
+		return nil, vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "tablet to avoid for reparent is not provided, cannot choose new primary")
+	}
+
+	var primaryCell string
+	if shardInfo.MasterAlias != nil {
+		primaryCell = shardInfo.MasterAlias.Cell
+	}
+
+	var (
+		searcher = topotools.NewMaxReplicationPositionSearcher(tmc, logger, waitReplicasTimeout)
+		wg       sync.WaitGroup
+	)
+
+	for _, tablet := range tabletMap {
+		switch {
+		case primaryCell != "" && tablet.Alias.Cell != primaryCell:
```

Comment:
This bothered me at first, but I think we looked at it once before and said that by default PRS won't choose a tablet in another cell, but you can always provide it a chosen tablet which will be respected. So I think this is fine.

Reply:
Yeah, this is a preservation of functionality. Maybe later we can make a vtctld startup flag to turn this off (I'd prefer to be able to PRS cross-cell without having to pick the tablet myself), but definitely far beyond the scope of this change.